### PR TITLE
boot: Provide our own EFI API headers

### PR DESCRIPTION
**Pull request by medhefgo**

### CODE REVIEW

**Comment by medhefgo** [comment]
I'm not too sure if using this style (functions defined inline in the structs) is better or if we should stick to the EFI spec style (lots of screaming typedefs for functions). Though, this style is fewer lines and yields nicer code completion when using clangd.

I'm also tempted to just use one large(r) efi.h header instead of splitting out the protocols. There isn't really any good way to split/group them, so a lot of them feel embarrassingly small.

If anyone has a preference for either of these, please speak up.

File: src/boot/efi/disk.h [hunk 1]
```diff
@@ -1,7 +1,5 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 #pragma once
 
-#include <efi.h>
-#include <uchar.h>
-
+#include "efi.h"
 EFI_STATUS disk_get_part_uuid(EFI_HANDLE *handle, char16_t uuid[static 37]);
```
Comment (yuwata) [inline src/boot/efi/disk.h]: Please insert an empty line after `#include`

File: src/boot/efi/efi-string.c [hunk 2]
```diff
@@ -1,14 +1,10 @@
...
-#include <stdbool.h>
-#include <stdint.h>
-#include <wchar.h>
-
 #include "efi-string.h"
 
 #if SD_BOOT
-#  include "missing_efi.h"
 #  include "util.h"
+#  include "proto-simple-text-io.h"
```
Comment (yuwata) [inline src/boot/efi/efi-string.c]: Please sort the headers.

File: src/boot/efi/shim.c [hunk 3]
```diff
@@ -8,10 +8,6 @@
  * https://github.com/mjg59/efitools
  */
 
-#include <efi.h>
-#include <efilib.h>
-
-#include "missing_efi.h"
 #include "util.h"
 #include "secure-boot.h"
 #include "shim.h"
```
Comment (yuwata) [inline src/boot/efi/shim.c]: Not fault in this commit, but please sort the headers.

File: src/boot/efi/part-discovery.c [hunk 4]
```diff
@@ -168,11 +168,6 @@ static EFI_STATUS try_gpt(
                 };
                 memcpy(ret_hd->Signature, &entry->UniquePartitionGUID, sizeof(ret_hd->Signature));
 
-                /* HARDDRIVE_DEVICE_PATH has padding, which at least OVMF does not like. */
-                SetDevicePathNodeLength(
-                                &ret_hd->Header,
-                                offsetof(HARDDRIVE_DEVICE_PATH, SignatureType) + sizeof(ret_hd->SignatureType));
```
Comment (yuwata) [inline src/boot/efi/part-discovery.c:159]: Why dropped? Is this change relevant to this commit?

**Review by yuwata — APPROVED** [review]
LGTM. Several minor comments and a question.

File: src/boot/efi/part-discovery.c (hunk 4 above)
Comment (medhefgo) [reply src/boot/efi/part-discovery.c:159]: The removed comment hints at it: our definition of the struct doesn't have the padding, so we can use sizeof directly when initializing it a few lines above.

Comment (yuwata) [reply src/boot/efi/part-discovery.c:159]: Thanks.

File: src/boot/efi/part-discovery.c [hunk 5]
```diff
@@ -37,14 +55,11 @@ static EFI_DEVICE_PATH *path_replace_hd(
         return ret;
 }
 
-static bool verify_gpt(union GptHeaderBuffer *gpt_header_buffer, EFI_LBA lba_expected) {
-        EFI_PARTITION_TABLE_HEADER *h;
+static bool verify_gpt(GptHeader *h, EFI_LBA lba_expected) {
```
Comment (poettering) [inline src/boot/efi/part-discovery.c]: maybe add a comment that the first argument is kinda const, but also not

File: src/boot/efi/util.c [hunk 6]
```diff
@@ -186,24 +186,25 @@ EFI_STATUS efivar_get_uint64_le(const EFI_GUID *vendor, const char16_t *name, ui
...
-        l = sizeof(char16_t *) * EFI_MAXIMUM_VARIABLE_SIZE;
-        buf = xmalloc(l);
+        size_t size = 0;
+        err = RT->GetVariable((char16_t *) name, (EFI_GUID *) vendor, NULL, &size, NULL);
+        if (err != EFI_BUFFER_TOO_SMALL)
+                return err;
 
-        err = RT->GetVariable((char16_t *) name, (EFI_GUID *) vendor, NULL, &l, buf);
+        _cleanup_free_ void *buf = xmalloc(size);
+        err = RT->GetVariable((char16_t *) name, (EFI_GUID *) vendor, NULL, &size, buf);
```
Comment (poettering) [inline src/boot/efi/util.c:200]: this smells as something that is going to trip up on shitty firmwares that lie about the size at first.

but let's try (and if we get reports this doesn't work, add a comment about it to the sources, to avoid the mess later on)

File: src/boot/efi/meson.build [hunk 7]
```diff
@@ -337,14 +335,27 @@ efi_headers = files(
...
+        'proto-dt-fixup.h',
+        'proto-file-io.h',
+        'proto-graphics-output.h',
+        'proto-load-file.h',
+        'proto-loaded-image.h',
+        'proto-rng.h',
+        'proto-security-arch.h',
+        'proto-shell-parameters.h',
+        'proto-simple-text-io.h',
+        'proto-tcg.h',
```
Comment (poettering) [inline src/boot/efi/meson.build]: hmm, I think it would make sense to move this into a subdir. These files are conceptually separate from out codebase, hence I think it makes sense to split them into their own dir, the same way we split out imported kernel headers and similar.

File: src/boot/efi/pe.c [hunk 8]
```diff
@@ -12,16 +8,16 @@
 #define MAX_SECTIONS 96
 
 #if defined(__i386__)
-#  define TARGET_MACHINE_TYPE EFI_IMAGE_MACHINE_IA32
-#  define TARGET_MACHINE_TYPE_COMPATIBILITY EFI_IMAGE_MACHINE_X64
+#  define TARGET_MACHINE_TYPE 0x014c
```
Comment (poettering) [inline src/boot/efi/pe.c]: should uses UINT16_C() i guess.

i.e. should be explicitly unsigned.

File: src/boot/efi/proto-simple-text-io.h [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+#define EFI_RIGHT_SHIFT_PRESSED   0x00000001
+#define EFI_LEFT_SHIFT_PRESSED    0x00000002
+#define EFI_RIGHT_CONTROL_PRESSED 0x00000004
+#define EFI_LEFT_CONTROL_PRESSED  0x00000008
+#define EFI_RIGHT_ALT_PRESSED     0x00000010
+#define EFI_LEFT_ALT_PRESSED      0x00000020
+#define EFI_RIGHT_LOGO_PRESSED    0x00000040
+#define EFI_LEFT_LOGO_PRESSED     0x00000080
+#define EFI_MENU_KEY_PRESSED      0x00000100
+#define EFI_SYS_REQ_PRESSED       0x00000200
```
Comment (poettering) [inline src/boot/efi/proto-simple-text-io.h]: as above, UINT32_C() (or at least suffix with U)

File: src/boot/efi/proto-file-io.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+#include "efi.h"
+
+#define EFI_SIMPLE_FILE_SYSTEM_PROTOCOL_GUID \
+        GUID_DEF(0x0964e5b22, 0x6459, 0x11d2, 0x8e, 0x39, 0x00, 0xa0, 0xc9, 0x69, 0x72, 0x3b)
+#define EFI_FILE_INFO_ID \
+        GUID_DEF(0x009576e92, 0x6d3f, 0x11d2, 0x8e, 0x39, 0x00, 0xa0, 0xc9, 0x69, 0x72, 0x3b)
+
+#define EFI_FILE_MODE_READ   0x0000000000000001
+#define EFI_FILE_MODE_WRITE  0x0000000000000002
+#define EFI_FILE_MODE_CREATE 0x8000000000000000
```
Comment (poettering) [inline src/boot/efi/proto-file-io.h]: so many zeroes....

I'd really make this all use UINT32_C() or suffix U because otherwise:

```c
typeof(EFI_FILE_MODE_READ) x = EFI_FILE_MODE_READ;
typeof(EFI_FILE_MODE_CREATE) y = EFI_FILE_MODE_CREATE;

assert(x < y);
```

would have an undefined result, since you'd compare a signed with an unsigned value, and `y` would change its sign if it would ne converted to int

File: src/boot/efi/proto-simple-text-io.h [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        SCAN_MUTE = 0x7F,
+        SCAN_VOLUME_UP = 0x80,
+        SCAN_VOLUME_DOWN = 0x81,
+        SCAN_BRIGHTNESS_UP = 0x100,
+        SCAN_BRIGHTNESS_DOWN = 0x101,
+        SCAN_SUSPEND = 0x102,
+        SCAN_HIBERNATE = 0x103,
+        SCAN_TOGGLE_DISPLAY = 0x104,
+        SCAN_RECOVERY = 0x105,
+        SCAN_EJECT = 0x106,
```
Comment (poettering) [inline src/boot/efi/proto-simple-text-io.h]: Zbigniew prefers it if these are aligned by the `=`...

File: src/boot/efi/part-discovery.c (hunk 5 above)
Comment (yuwata) [reply src/boot/efi/part-discovery.c]: Not addressed yet.

**Review by yuwata — APPROVED** [review]
LGTM. But one suggestion by @poettering not addressed yet.

**Comment by keszybz** [comment]
jammy-amd64 and jammy-s390x are stuck, but the rpm builds on all architectures succeeded, so this should be fine.
